src/openslide-decode-dicom.c

```diff
@@ -33,73 +33,89 @@ void _openslide_dicom_propagate_error(GError **err, DcmError *dcm_error) {
   dcm_error_clear(&dcm_error);
 }
 
+// implements DcmIO
+struct _dicom_io {
+    DcmIOMethods *methods;
+    struct _openslide_file *file;
+};
+
 static void propagate_gerror(DcmError **dcm_error, GError *err) {
   dcm_error_set(dcm_error, DCM_ERROR_CODE_INVALID,
                 g_quark_to_string(err->domain),
                 "%s", err->message);
   g_error_free(err);
 }
 
-static void *vfs_open(DcmError **dcm_error, void *client) {
+static DcmIO *vfs_open(DcmError **dcm_error, void *client) {
   const char *filename = (const char *) client;
+  struct _dicom_io *dio = g_new(struct _dicom_io, 1);
 
   GError *err = NULL;
-  struct _openslide_file *file = _openslide_fopen(filename, &err);
-  if (!file) {
+  dio->file = _openslide_fopen(filename, &err);
+  if (!dio->file) {
     propagate_gerror(dcm_error, err);
     return NULL;
```

Comment:
Leaks `dio`.  For consistency, let's put a `g_autofree` on `dio` and use `g_steal_pointer()` on return.

Reply:
Actually, `g_autofree` isn't right either, since there's a field that would be leaked if we added any additional fallible initialization.  It's probably not worth a dedicated cleanup function for one call site, so I guess we just do the cleanup inline for now.

Reply:
Ooop, good catch. I added a `g_free()` on the error path.